lib/cfg.js

```diff
@@ -0,0 +1,368 @@
+
+                isBasicBlockEnd: function(x) { return x[0] == ".";} ,
+
+                getInstructionType:function(inst) {
+                                            inst = inst.trim();
+                                            if (inst === "") return -1;
+                                            else if(inst.includes("jmp")) return 0;
+                                            else if(inst[0] === 'j') return 1;
+                                            else if(!inst.includes("ret")) return 2;
+                                            else return 3;
+                                          },
+
+                extractNodeName:function(inst) {
+                                     var name = inst.match(/.L\d+/);
+                                     return name + ":";
+                                  },
+
+                isJmpInstruction: function(x) { var trimed = x.trim(); return ( trimed[0] == 'j');}
+             };
+
+
+
+
+ControlFlowGraph.prototype.splitToFunctions = function (asmArr, isEnd) {
+    if(asmArr.length === 0)   return [];
+    var result = [];
+    var first = 0;
+    var last = asmArr.length;
+    var fnRange = {start:first, end:null};
+    ++first;
+    while(first != last) {
+        if(isEnd(asmArr[first].text)) {
+            fnRange.end = first;
+            result.push(JSON.parse(JSON.stringify(fnRange)));
+            fnRange.start = first;
+        }
+        ++first;
+    }
+
+    fnRange.end = last;
+    result.push(fnRange);//possiblitiy of bug?! because didn't clone(JSON.parse(JSON.stringify(...))
+    return result;
+};
+
+ControlFlowGraph.prototype.splitToBasicBlocks = function (asmArr,  range, isEnd, isJmp) {
+    var first = range.start;
+    var last = range.end;
+    if(first == last) return [];
+    ++first;
+
+    var rangeBb = {nameId:"start" , start:first , end:null, actionPos:[]};
+    var result = [];
+
+    var resetRangeWith = function(rangeBb,  nameId, start) {
+            rangeBb.nameId = nameId;//.L1:
+            rangeBb.start = start;//after .L1:
+            rangeBb.actionPos = [];
+    };
+
+    while(first != last) {
+        var inst = asmArr[first].text;
+        if(isEnd(inst)) {
+            rangeBb.end = first;
+            result.push(JSON.parse(JSON.stringify(rangeBb)));
+            ++first;//risk of a bug.
+            //inst is expected to be .L*: where * in 1,2,...
+            resetRangeWith(rangeBb, inst, first);
+        }
+        else if(isJmp(inst)) {
+                rangeBb.actionPos.push(first);
+        }
+        ++first;
+    }
+
+    rangeBb.end = last;
+    result.push(JSON.parse(JSON.stringify(rangeBb)));
+    return result;
+};
+
+ControlFlowGraph.prototype.splitToCanonicalBasicBlock = function (basicBlock) {
+    var actionPos = basicBlock.actionPos;
+    var actPosSz = actionPos.length;
+    if (actionPos[actPosSz-1] + 1 == basicBlock.end){
+        --actPosSz;
+    }
+
+    if(actPosSz === 0)
```

Comment:
I wonder if there's a nice way to write the code so it doesn't have to special case 0 and 1. I can't see from a quick look but usually it's a bad code smell to have to handle these kinds of special cases separately.

Reply:
is it possible to defer to another time?

Reply:
Yes. Let's get the tests in, and UI fix in, then we can always refactor.